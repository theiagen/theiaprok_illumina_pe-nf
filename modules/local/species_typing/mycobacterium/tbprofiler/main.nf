process TBPROFILER {
    tag "$meta.id"
    label 'process_medium'

    container "us-docker.pkg.dev/general-theiagen/staphb/tbprofiler:6.6.3"

    input:
    tuple val(meta), path(reads)
    val ont_data // set to false if not ONT data, Theiaprok_Illumina_PE/SE/FASTA

    output:
    tuple val(meta), path("results/*.results.csv"), emit: csv
    tuple val(meta), path("results/*.results.txt"), emit: txt
    tuple val(meta), path("results/*.results.json"), emit: json
    tuple val(meta), path("bam/*.bam"), emit: bam
    tuple val(meta), path("bam/*.bam.bai"), emit: bai
    tuple val(meta), path("vcf/*.targets.csq.merged.vcf"), optional: true, emit: vcf
    tuple val(meta), path("VERSION"), emit: version
    tuple val(meta), path("MAIN_LINEAGE"), emit: main_lineage
    tuple val(meta), path("SUB_LINEAGE"), emit: sub_lineage
    tuple val(meta), path("DR_TYPE"), emit: dr_type
    tuple val(meta), path("NUM_DR_VARIANTS"), emit: num_dr_variants
    tuple val(meta), path("NUM_OTHER_VARIANTS"), emit: num_other_variants
    tuple val(meta), path("RESISTANCE_GENES"), emit: resistance_genes
    tuple val(meta), path("MEDIAN_DEPTH"), emit: median_depth
    tuple val(meta), path("PCT_READS_MAPPED"), emit: pct_reads_mapped

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def read1 = reads[0]
    def read2 = reads.size() > 1 ? reads[1] : null
    def is_ont = ont_data ? "--platform nanopore": ""
    def variant_caller = params.tbprofiler_variant_caller ?: "gatk"
    def variant_calling_params = params.tbprofiler_variant_calling_params ?: ""
    def min_depth = params.tbprofiler_min_depth ?: 10
    def min_af = params.tbprofiler_min_af ?: 0.1
    def tbprofiler_custom_db = params.tbprofiler_custom_db ?: ""
    def tbprofiler_run_custom_db = params.tbprofiler_run_custom_db ?: false
    def tbprofiler_run_cdph_db = params.tbprofiler_run_cdph_db ?: false
    def tbprofiler_mapper = params.tbprofiler_mapper ?: "bwa"
    def args = task.ext.args ?: ""

    """
    # Print and save version
    tb-profiler version > VERSION && sed -i -e 's/TBProfiler version //' VERSION && sed -n -i '\$p' VERSION
    
    # check if file is non existant or non empty
    if [ -z "${read2}" ] || [ ! -s "${read2}" ]; then
      INPUT_READS="-1 ${read1}"
    else
      INPUT_READS="-1 ${read1} -2 ${read2}"
    fi

    # check if new database file is provided and not empty
    TBDB=""
    if ${tbprofiler_run_custom_db}; then
      if [ ! -s ${tbprofiler_custom_db} ]; then
        echo "Custom database file is empty"
      else
        echo "Found new database file ${tbprofiler_custom_db}"
        prefix=\$(basename "${tbprofiler_custom_db}" | sed 's/\\.tar\\.gz\$//')
        tar xfv ${tbprofiler_custom_db}
        
        tb-profiler load_library ./"\$prefix"/"\$prefix"

        TBDB="--db \$prefix"
      fi
    elif ${tbprofiler_run_cdph_db}; then
      tb-profiler update_tbdb --branch CaliforniaDPH
      TBDB="--db CaliforniaDPH"
    fi

    # Run tb-profiler on the input reads with prefix prefix
    tb-profiler profile \\
      ${read1} ${read2} \\
      --prefix ${prefix} \\
      --mapper ${tbprofiler_mapper} \\
      --caller ${variant_caller} \\
      --calling_params "${variant_calling_params}" \\
      --depth ${min_depth} \\
      --af ${min_af} \\
      --threads ${task.cpus} \\
      --csv --txt \\
      ${is_ont} \\
      ${args} \\
      \${TBDB}

    # Collate results
    tb-profiler collate --prefix ${prefix}
    ls -la ./*/*
    # merge all vcf files if multiple are present
    bcftools index ./vcf/*.bcf
    bcftools index ./vcf/*.gz
    bcftools merge --force-samples ./vcf/*.bcf ./vcf/*.gz > ./vcf/${prefix}.targets.csq.merged.vcf
    
    parse_tbprofiler.py --input_file results/${prefix}.results.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        tbprofiler: \$(VERSION)
    END_VERSIONS
    """
    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    # Create empty files for stub
    tb-profiler version > VERSION && sed -i -e 's/TBProfiler version //' VERSION && sed -n -i '\$p' VERSION
    mkdir results bam vcf
    touch results/${prefix}.results.csv
    touch results/${prefix}.results.txt
    touch results/${prefix}.results.json
    touch bam/${prefix}.bam
    touch bam/${prefix}.bam.bai
    touch vcf/${prefix}.targets.csq.merged.vcf
    echo "na" > MAIN_LINEAGE
    echo "na" > SUB_LINEAGE
    echo "na" > DR_TYPE
    echo "na" > NUM_DR_VARIANTS
    echo "na" > NUM_OTHER_VARIANTS
    echo "na" > RESISTANCE_GENES
    echo "na" > MEDIAN_DEPTH
    echo "na" > PCT_READS_MAPPED
    

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        tbprofiler: \$(VERSION)
    END_VERSIONS
    """
}