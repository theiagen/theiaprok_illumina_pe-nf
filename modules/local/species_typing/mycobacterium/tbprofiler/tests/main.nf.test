nextflow_process {
    name "Test Process TBPROFILER"
    script "../main.nf"
    process "TBPROFILER"
    
    test("tbprofiler test") {
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    [
                    file(params.modules_local_testdata_base_path + "SRX11523932_1.fastq", checkIfExists: true),
                    file(params.modules_local_testdata_base_path + "SRX11523932_2.fastq", checkIfExists: true)
                    ]
                ])
                input[1] = false
                params.tbprofiler_variant_caller = "gatk"
                params.tbprofiler_variant_calling_params = ""
                params.tbprofiler_min_depth = 10
                params.tbprofiler_min_af = 0.1
                params.tbprofiler_custom_db = ""
                params.tbprofiler_run_custom_db = false
                params.tbprofiler_run_cdph_db = false
                params.tbprofiler_mapper = "bwa"
                params.tbprofiler_args = ""
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.cleaned_reads,
                    process.out.versions
                    ).match() }
            )
        }
    }
    
    test("tbprofiler - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    [
                    file(params.modules_local_testdata_base_path + "SRX11523932_1.fastq", checkIfExists: true),
                    file(params.modules_local_testdata_base_path + "SRX11523932_2.fastq", checkIfExists: true)
                    ]
                ])
                input[1] = false
                params.tbprofiler_variant_caller = "gatk"
                params.tbprofiler_variant_calling_params = ""
                params.tbprofiler_min_depth = 10
                params.tbprofiler_min_af = 0.1
                params.tbprofiler_custom_db = ""
                params.tbprofiler_run_custom_db = false
                params.tbprofiler_run_cdph_db = false
                params.tbprofiler_mapper = "bwa"
                params.tbprofiler_args = ""
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}