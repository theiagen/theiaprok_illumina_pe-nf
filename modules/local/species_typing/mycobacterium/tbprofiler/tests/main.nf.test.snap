{
    "tbprofiler - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.results.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.results.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "10": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "NUM_DR_VARIANTS:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "11": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "NUM_OTHER_VARIANTS:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "12": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "RESISTANCE_GENES:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "13": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "MEDIAN_DEPTH:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "14": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "PCT_READS_MAPPED:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.results.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.bam.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.targets.csq.merged.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "VERSION:md5,b32d1e9ad34bb4c363d507c86077bc58"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "MAIN_LINEAGE:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "SUB_LINEAGE:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "9": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "DR_TYPE:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "bai": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.bam.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "bam": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "csv": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.results.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "dr_type": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "DR_TYPE:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "json": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.results.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "main_lineage": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "MAIN_LINEAGE:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "median_depth": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "MEDIAN_DEPTH:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "num_dr_variants": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "NUM_DR_VARIANTS:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "num_other_variants": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "NUM_OTHER_VARIANTS:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "pct_reads_mapped": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "PCT_READS_MAPPED:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "resistance_genes": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "RESISTANCE_GENES:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "sub_lineage": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "SUB_LINEAGE:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "txt": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.results.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.targets.csq.merged.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "version": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "VERSION:md5,b32d1e9ad34bb4c363d507c86077bc58"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.8"
        },
        "timestamp": "2025-06-25T16:57:07.856344048"
    },
    "tbprofiler test": {
        "content": [
            null,
            null
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.8"
        },
        "timestamp": "2025-06-25T16:57:02.73101476"
    }
}